gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java

```diff
@@ -231,13 +326,136 @@ private ManagedChannel getNextChannel() {
+      this.channel = channel;
+    }
+
+    /**
+     * Try to increment the outstanding RPC count. The method will return false if the channel is
+     * closing and the caller should pick a different channel. If the method returned true, the
+     * channel has been successfully retained and it is the responsibility of the caller to release
+     * it.
+     */
+    private boolean retain() {
+      // register desire to start RPC
+      outstandingRpcs.incrementAndGet();
+
+      // abort if the channel is closing
+      if (shutdownRequested.get()) {
+        release();
+        return false;
+      }
+      return true;
+    }
+
+    /**
+     * Notify the channel that the number of outstanding RPCs has decreased. If shutdown has been
+     * previously requested, this method will shutdown the channel if its the last outstanding RPC.
+     */
+    private void release() {
+      int newCount = outstandingRpcs.decrementAndGet();
+      if (newCount < 0) {
+        LOG.log(Level.SEVERE, "Reference count is negative!: " + newCount);
+      }
+
+      if (newCount == 0 && shutdownRequested.get()) {
+        shutdown();
+      }
+    }
+
+    /**
+     * Request a shutdown. The actual shutdown will be delayed until there are no more outstanding
+     * RPCs.
+     */
+    private void requestShutdown() {
+      shutdownRequested.set(true);
+      if (outstandingRpcs.get() == 0) {
+        shutdown();
+      }
+    }
+
+    /** Ensure that shutdown is only called once. */
+    private void shutdown() {
+      if (shutdownInitiated.compareAndSet(false, true)) {
+        channel.shutdown();
+      }
+    }
+  }
+
+  /** Thin wrapper to ensure that new calls are properly reference counted. */
+  private class AffinityChannel extends Channel {
+    private final int affinity;
+
+    public AffinityChannel(int affinity) {
+      this.affinity = affinity;
+    }
+
+    @Override
+    public String authority() {
+      return authority;
+    }
+
+    @Override
+    public <RequestT, ResponseT> ClientCall<RequestT, ResponseT> newCall(
+        MethodDescriptor<RequestT, ResponseT> methodDescriptor, CallOptions callOptions) {
+
+      Entry entry = getRetainedEntry(affinity);
+
+      return new ReleasingClientCall<>(entry.channel.newCall(methodDescriptor, callOptions), entry);
+    }
+  }
+
+  /** ClientCall wrapper that makes sure to decrement the outstanding RPC count on completion. */
+  static class ReleasingClientCall<ReqT, RespT> extends SimpleForwardingClientCall<ReqT, RespT> {
+    final Entry entry;
+
+    public ReleasingClientCall(ClientCall<ReqT, RespT> delegate, Entry entry) {
+      super(delegate);
+      this.entry = entry;
+    }
+
+    @Override
+    public void start(Listener<RespT> responseListener, Metadata headers) {
+      try {
+        super.start(
+            new SimpleForwardingClientCallListener<RespT>(responseListener) {
+              @Override
+              public void onClose(Status status, Metadata trailers) {
```

Comment:
Is it guaranteed that `onClose()` is always called (e.g., even in error situations)?

Reply:
Yes, if the call starts, then onClose will be called